### ISSUE TITLE
Support vfork for sys_clone3  in Asterinas

### ISSUE THREAD

**Issue by grief8** [asterinas/asterinas#1485]
#### Description:
Currently, the Asterinas project does not fully support the `vfork` system call, which is represented by the `CLONE_VFORK` flag in the `clone` system call. The `vfork` system call is used to create a new process that shares the same memory space as the parent process until the child process either calls `exec` or exits. This is particularly useful for performance reasons in scenarios where the child process will immediately replace its memory image with a new program.

#### Proposed Changes:
1. **Implementation of `CLONE_VFORK` Flag Handling**:
   - Modify the `clone_child_task` and `clone_child_process` functions to handle the `CLONE_VFORK` flag appropriately.
   - Ensure that the child process shares the same memory space as the parent until it calls `exec` or exits.

2. **Parent Process Blocking**:
   - Implement logic to block the parent process until the child process calls `exec` or exits when `CLONE_VFORK` is set.

3. **Memory Management**:
   - Ensure that the memory management logic correctly handles the shared memory space between the parent and child processes during the `vfork` operation.

4. **Testing**:
   - Add unit tests to verify the correct behavior of `vfork`, including scenarios where the child process calls `exec` and exits.

#### Example Code Snippet:
```rust
rust
fn clone_child_task(
ctx: &Context,
parent_context: &UserContext,
clone_args: CloneArgs,
) -> Result<Arc<Task>> {
// ... existing code ...
if clone_flags.contains(CloneFlags::CLONE_VFORK) {
// Block the parent process until the child calls exec or exits
// Implement the necessary logic here
}
// ... existing code ...
}
```

#### Benefits:
- **Performance Improvement**: `vfork` can significantly improve performance in scenarios where the child process will immediately replace its memory image with a new program.
- **Feature Parity**: Supporting `vfork` brings Asterinas closer to feature parity with traditional Unix-like systems.

#### Conclusion:
Adding support for `vfork` (CLONE_VFORK) in Asterinas will enhance the project's capabilities and provide a more efficient way to create new processes in certain scenarios. This feature request aims to address the current lack of support for `vfork` and proposes a plan to implement it.

**Comment by grief8** [comment]
@cchanging would you shed some light on this issue?

**Comment by cchanging** [comment]
> @cchanging would you shed some light on this issue?

I'll take a look and try to implement it correctly.

**Comment by le-monde-bleu** [comment]
The busybox `init` process also requires the support of `vfork` syscall. As mentioned in #1429 , change the `init_args` into `["init"]` in `asterinas/OSDK.toml` (so as to relocate the booting process to start with the booting process of busybox `init`), and add a `/test/build/initramfs/etc/inittab` file to set the `init` process like

> `::respawn:/bin/sh`

This should boot the system and switch to shell. However, it reports

> `can't fork`

due to the deficiency of `vfork` syscall. See the `init` code of busybox in

(https://github.com/brgl/busybox/blob/abbf17abccbf832365d9acf1c280369ba7d5f8b2/init/init.c#L483)

What could I possibly do to help implement the `vfork` syscall?

**Comment by cchanging** [comment]
@le-monde-bleu Thank you for your provided information. If the absence of `vfork` blocks your work, then we can indeed prioritize the implementation of this feature. Currently, to implement `vfork`, we need to add two main functionalities on top of the existing `fork` logic: one is to block the parent during vfork until the child exits or calls `execute` syscall, and the other is to create a `ProcessVm` independent of the parent when the child calls `execve`. 

I need to confirm with you: what is the main reason for using `vfork` in the program code you mentioned in your case? Could it directly uses fork as a dummy? Or does it only needs to block the parent? Clarifying this point should allow us to make phased modifications that can help you advance your work more quickly.

**Comment by le-monde-bleu** [comment]
@cchanging Thank you for your attention! I am trying to simulate the booting process of an embedded Linux system (as referred to in [https://heygoda.github.io/Hello-Embedded-Linux/%E7%B3%BB%E7%BB%9F%E5%88%9D%E5%A7%8B%E5%8C%96.html](https://github.com/asterinas/asterinas/issues/url)). Asterinas implements the `init` process of busybox as the booting process, while `init` may trigger the `vfork` syscall. As I mentioned

>  due to the deficiency of vfork syscall. See the init code of busybox in (https://github.com/brgl/busybox/blob/abbf17abccbf832365d9acf1c280369ba7d5f8b2/init/init.c#L483)

It may be difficult to change the code of `init`, so it is more reasonable to implement `vfork`.

> Could it directly uses fork as a dummy?

I am not quite sure about this, but I quess it may be logically incorrect.

May I take over the work of implementing `vfork`, if it is not that urgent (say should be done in one day or two).

**Comment by cchanging** [comment]
> May I take over the work of implementing vfork, if it is not that urgent (say should be done in one day or two).

Of course! Feel free to bring up any questions of implementation for discussion.

Note that in the busybox code you provided, it seems that the reason for using `vfork` in the second branch might be due to the lack of support for certain features, which would make copying the page table slow (I guess). Therefore, using vfork reduces one memory copy operation. You can try to use `fork` to implement `vfork` directly first to see if there is only a performance difference.

**Comment by grief8** [comment]
close as #1484
